File: server/combat/arithmetic.go
package combat

import (
	"errors"
	"math"
)

var ErrInvalidArithmetic = errors.New("invalid combat arithmetic")

type Temperament string

const (
	Lazy    Temperament = "lazy"
	Playful Temperament = "playful"
	Curious Temperament = "curious"
	Sassy   Temperament = "sassy"
	Shy     Temperament = "shy"
	Chaotic Temperament = "chaotic"
)

type ChartResult int8

const (
	Disadvantage ChartResult = -1
	Neutral      ChartResult = 0
	Advantage    ChartResult = 1
)

var temperamentOrder = [...]Temperament{Lazy, Playful, Curious, Sassy, Shy, Chaotic}

func Chart(attacker, defender Temperament) (ChartResult, error) {
	attackerIndex, attackerOK := temperamentIndex(attacker)
	defenderIndex, defenderOK := temperamentIndex(defender)
	if !attackerOK || !defenderOK {
		return Neutral, ErrInvalidArithmetic
	}
	delta := (defenderIndex - attackerIndex + len(temperamentOrder)) % len(temperamentOrder)
	switch delta {
	case 1, 2:
		return Advantage, nil
	case 4, 5:
		return Disadvantage, nil
	default:
		return Neutral, nil
	}
}

func Damage(basePower, attackerATK int32, chart ChartResult, critical bool) (int32, error) {
	if basePower < 0 || attackerATK < 0 || chart < Disadvantage || chart > Advantage {
		return 0, ErrInvalidArithmetic
	}
	if basePower == 0 {
		return 0, nil
	}
	value := int64(basePower)
	value = floorRatio(value, int64(attackerATK), 64)
	switch chart {
	case Advantage:
		value = floorRatio(value, 13, 10)
	case Disadvantage:
		value = floorRatio(value, 10, 13)
	}
	if critical {
		value = floorRatio(value, 3, 2)
	}
	if value < 1 {
		value = 1
	}
	return SaturateInt32(value), nil
}

func SaturateInt32(value int64) int32 {
	if value > math.MaxInt32 {
		return math.MaxInt32
	}
	if value < math.MinInt32 {
		return math.MinInt32
	}
	return int32(value)
}

func Clamp(value, minimum, maximum int64) (int32, error) {
	if minimum < 0 || maximum < minimum || maximum > math.MaxInt32 {
		return 0, ErrInvalidArithmetic
	}
	if value < minimum {
		value = minimum
	}
	if value > maximum {
		value = maximum
	}
	return int32(value), nil
}

func floorRatio(value, numerator, denominator int64) int64 {
	if value <= 0 || numerator <= 0 || denominator <= 0 {
		return 0
	}
	return value * numerator / denominator
}

func temperamentIndex(value Temperament) (int, bool) {
	for index, candidate := range temperamentOrder {
		if value == candidate {
			return index, true
		}
	}
	return 0, false
}

File: client/src/combat/arithmetic.ts
import { idiv } from "../integer";

export type Temperament = "lazy" | "playful" | "curious" | "sassy" | "shy" | "chaotic";
export type ChartResult = -1 | 0 | 1;

const order: readonly Temperament[] = Object.freeze(["lazy", "playful", "curious", "sassy", "shy", "chaotic"]);
const INT32_MAX = 2_147_483_647n;
const INT32_MIN = -2_147_483_648n;

export function chart(attacker: Temperament, defender: Temperament): ChartResult {
  const attackerIndex = order.indexOf(attacker);
  const defenderIndex = order.indexOf(defender);
  if (attackerIndex < 0 || defenderIndex < 0) throw new RangeError("unknown temperament");
  const delta = (defenderIndex - attackerIndex + order.length) % order.length;
  if (delta === 1 || delta === 2) return 1;
  if (delta === 4 || delta === 5) return -1;
  return 0;
}

export function damage(basePower: number, attackerAtk: number, result: ChartResult, critical: boolean): number {
  int32(basePower, "base power"); int32(attackerAtk, "attacker atk");
  if (basePower < 0 || attackerAtk < 0 || (result !== -1 && result !== 0 && result !== 1)) throw new RangeError("invalid damage input");
  if (basePower === 0) return 0;
  let value = idiv(BigInt(basePower) * BigInt(attackerAtk), 64n);
  if (result === 1) value = idiv(value * 13n, 10n);
  if (result === -1) value = idiv(value * 10n, 13n);
  if (critical) value = idiv(value * 3n, 2n);
  if (value < 1n) value = 1n;
  return saturateInt32(value);
}

export function saturateInt32(value: bigint): number {
  if (value > INT32_MAX) return Number(INT32_MAX);
  if (value < INT32_MIN) return Number(INT32_MIN);
  return Number(value);
}

export function clamp(value: bigint, minimum: number, maximum: number): number {
  int32(minimum, "minimum"); int32(maximum, "maximum");
  if (minimum < 0 || maximum < minimum) throw new RangeError("invalid clamp bounds");
  if (value < BigInt(minimum)) value = BigInt(minimum);
  if (value > BigInt(maximum)) value = BigInt(maximum);
  return Number(value);
}

function int32(value: number, label: string): void {
  if (!Number.isInteger(value) || value < Number(INT32_MIN) || value > Number(INT32_MAX)) throw new RangeError(`${label} outside int32`);
}
